cstrml/staking/src/lib.rs

```diff
@@ -1544,59 +1649,86 @@ impl<T: Trait> Module<T> {
                 // we should using `FILO` rule to calculate others
-                if remains > 0 && votes > 0 {
+                if remains > 0 {
                     let g_real_votes = remains.min(votes);
                     let g_vote_stakes = to_balance(g_real_votes);
 
-                    // a. preparing new_guarantors for `Validators`, others for `Stakers`
+                    // a. preparing new_guarantors for `Validators`
                     new_guarantors.push(guarantor.clone());
-                    others.push(IndividualExposure {
-                        who: guarantor.clone(),
-                        value: g_vote_stakes,
-                    });
                     // b. update remains and guarantors votes
                     remains -= g_real_votes;
                     v_guarantors_votes += g_real_votes;
+                    
+                    // c. UPDATE EDGE: (maybe) update GuaranteeRel and Nominations.total
+                    if g_real_votes < votes {
+                        if let Some(mut nominations) = Self::guarantors(&guarantor) {	
+                            nominations.total -= to_balance(votes - g_real_votes);
+                            <Guarantors<T>>::insert(&guarantor, nominations);
+                        }
+                        records.insert(index, g_vote_stakes);
+                        <GuaranteeRel<T>>::insert(&guarantor, &v_stash, records);
+                    }
 
-                    // c. UPDATE EDGE: (maybe) update GuaranteeRel
-                    <GuaranteeRel<T>>::insert(&guarantor, &v_stash, g_vote_stakes);
-
-                // 2. There has no credit for later guarantors
-                // or guarantor already chilled, update `Guarantors`
-                // and remove `GuaranteeRel`
+                // There has no credit for later guarantors
                 } else {
-                    // a. UPDATE NODE: update `Guarantors`
-                    // TODO: this can merge into II's logic, that will be more efficient
-                    if let Some(mut nominations) = Self::guarantors(guarantor) {
-                        nominations.targets.remove_item(v_stash);
-                        <Guarantors<T>>::insert(guarantor, nominations);
+                    records.remove(&index);
+                    let is_empty = records.len() == 0;
+                    if is_empty {
+                        <GuaranteeRel<T>>::remove(&guarantor, &v_stash);
+                    } else {
+                        <GuaranteeRel<T>>::insert(&guarantor, &v_stash, records);
+                    }
+                    // Update Nominations
+                    if let Some(mut nominations) = Self::guarantors(&guarantor) {	
+                        nominations.total -= to_balance(votes);
+                        if is_empty {
+                            nominations.targets.remove_item(v_stash);	
+                        }
+                        <Guarantors<T>>::insert(&guarantor, nominations);	
```

Comment:
If we do not use `records`, we can simplify this logic into:
```rust
else {
    // a. Remove edge
    <GuaranteeRel<T>>::mutate(xxx) // if index is 0, remove whole edge and let removed = true
    // b. Update Guarantors
    // if removed is true, remove from targets
}
```

Reply:
moved records into else section.

Reply:
This is clearer to me